File: CustomApplications/CSharp/GraphicsHowTo/PositionOrientationProvider.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using AGI.STKUtil;
using AGI.STKObjects;

namespace GraphicsHowTo
{
    public class PositionOrientationProvider
    {
        private const string Separator = "    ";

        public PositionOrientationProvider(string filename, AgStkObjectRoot root)
        {
            m_Dates = new List<double>();
            m_Positions = new List<Array>();
            m_Orientations = new List<Array>();
            m_root = root;

            using (StreamReader sr = new StreamReader(filename))
            {
                while (sr.Peek() >= 0)
                {
                    string[] sEntries = sr.ReadLine().Replace(Separator, ",").Split(',');
                    m_Dates.Add(double.Parse(root.ConversionUtility.NewDate("UTCG", sEntries[0]).Format("epSec")));

                    double x = Double.Parse(sEntries[1], CultureInfo.InvariantCulture);
                    double y = Double.Parse(sEntries[2], CultureInfo.InvariantCulture);
                    double z = Double.Parse(sEntries[3], CultureInfo.InvariantCulture);
                    Array pos = new object[] { x, y, z };
                    m_Positions.Add(pos);

                    x = Double.Parse(sEntries[4], CultureInfo.InvariantCulture);
                    y = Double.Parse(sEntries[5], CultureInfo.InvariantCulture);
                    z = Double.Parse(sEntries[6], CultureInfo.InvariantCulture);
                    double w = Double.Parse(sEntries[7], CultureInfo.InvariantCulture);
                    Array orientation = new object[] { x, y, z, w };
                    m_Orientations.Add(orientation);
                }
            }
        }

        public IList<double> Dates
        {
            get { return m_Dates; }
        }
        public IList<Array> Positions
        {
            get { return m_Positions; }
        }
        public IList<Array> Orientations
        {
            get { return m_Orientations; }
        }

        public int FindIndexOfClosestTime(double searchTime, int startIndex, int searchLength)
        {
            // Find the midpoint of the length
            int midpoint = startIndex + (searchLength / 2);

            // Base cases
            if (m_Dates[startIndex] == searchTime || searchLength == 1)
            {
                return startIndex;
            }
            if (searchLength == 2)
            {
                double diff1 = m_Dates[startIndex] - searchTime;
                double diff2 = m_Dates[startIndex + 1] - searchTime;

                if (Math.Abs(diff1) < Math.Abs(diff2))
                {
                    return startIndex;
                }
                else // Note: error on the larger time if equal
                {
                    return startIndex + 1;
                }
            }
            if (m_Dates[midpoint] == searchTime)
            {
                return midpoint;
            }

            // Normal case: binary search
            if (searchTime < m_Dates[midpoint])
            {
                return FindIndexOfClosestTime(searchTime, startIndex, midpoint - startIndex);
            }
            else
            {
                return FindIndexOfClosestTime(searchTime, midpoint + 1, startIndex + searchLength - (midpoint + 1));
            }
        }


        private readonly IList<double> m_Dates;
        private readonly IList<Array> m_Positions;
        private readonly IList<Array> m_Orientations;
        private readonly AgStkObjectRoot m_root;
    }
}

File: CustomApplications/VB.Net/GraphicsHowTo/PositionOrientationProvider.vb
Imports System.Collections.Generic
Imports System.Globalization
Imports System.IO
Imports AGI.STKUtil
Imports AGI.STKObjects

Public Class PositionOrientationProvider
	Private Const Separator As String = "    "

	Public Sub New(filename As String, root As AgStkObjectRoot)
		m_Dates = New List(Of Double)()
		m_Positions = New List(Of Array)()
		m_Orientations = New List(Of Array)()
		m_root = root

		Using sr As New StreamReader(filename)
			While sr.Peek() >= 0
				Dim sEntries As String() = sr.ReadLine().Replace(Separator, ",").Split(","C)
				m_Dates.Add(Double.Parse(root.ConversionUtility.NewDate("UTCG", sEntries(0)).Format("epSec")))

                Dim x As Double = Double.Parse(sEntries(1), CultureInfo.InvariantCulture)
                Dim y As Double = Double.Parse(sEntries(2), CultureInfo.InvariantCulture)
                Dim z As Double = Double.Parse(sEntries(3), CultureInfo.InvariantCulture)
				Dim pos As Array = New Object() {x, y, z}
				m_Positions.Add(pos)

                x = Double.Parse(sEntries(4), CultureInfo.InvariantCulture)
                y = Double.Parse(sEntries(5), CultureInfo.InvariantCulture)
                z = Double.Parse(sEntries(6), CultureInfo.InvariantCulture)
                Dim w As Double = Double.Parse(sEntries(7), CultureInfo.InvariantCulture)
				Dim orientation As Array = New Object() {x, y, z, w}
				m_Orientations.Add(orientation)
			End While
		End Using
	End Sub

	Public ReadOnly Property Dates() As IList(Of Double)
		Get
			Return m_Dates
		End Get
	End Property
	Public ReadOnly Property Positions() As IList(Of Array)
		Get
			Return m_Positions
		End Get
	End Property
	Public ReadOnly Property Orientations() As IList(Of Array)
		Get
			Return m_Orientations
		End Get
	End Property

	Public Function FindIndexOfClosestTime(searchTime As Double, startIndex As Integer, searchLength As Integer) As Integer
		' Find the midpoint of the length
		Dim midpoint As Integer = startIndex + (searchLength \ 2)

		' Base cases
		If m_Dates(startIndex) = searchTime OrElse searchLength = 1 Then
			Return startIndex
		End If
		If searchLength = 2 Then
			Dim diff1 As Double = m_Dates(startIndex) - searchTime
			Dim diff2 As Double = m_Dates(startIndex + 1) - searchTime

			If Math.Abs(diff1) < Math.Abs(diff2) Then
				Return startIndex
			Else
				' Note: error on the larger time if equal
				Return startIndex + 1
			End If
		End If
		If m_Dates(midpoint) = searchTime Then
			Return midpoint
		End If

		' Normal case: binary search
		If searchTime < m_Dates(midpoint) Then
			Return FindIndexOfClosestTime(searchTime, startIndex, midpoint - startIndex)
		Else
			Return FindIndexOfClosestTime(searchTime, midpoint + 1, startIndex + searchLength - (midpoint + 1))
		End If
	End Function


	Private ReadOnly m_Dates As IList(Of Double)
	Private ReadOnly m_Positions As IList(Of Array)
	Private ReadOnly m_Orientations As IList(Of Array)
	Private ReadOnly m_root As AgStkObjectRoot
End Class
